winit-wayland/src/types/xdg_toplevel_icon_manager.rs

```diff
@@ -0,0 +1,112 @@
+//! Handling of xdg toplevel icon manager, which is used for icon setting requests.
+
+use sctk::globals::GlobalData;
+use sctk::shm::slot::{Buffer, SlotPool};
+use tracing::warn;
+use wayland_client::globals::{BindError, GlobalList};
+use wayland_client::protocol::wl_shm::Format;
+use wayland_client::{delegate_dispatch, Connection, Dispatch, Proxy, QueueHandle};
+use wayland_protocols::xdg::toplevel_icon::v1::client::xdg_toplevel_icon_manager_v1::XdgToplevelIconManagerV1;
+use wayland_protocols::xdg::toplevel_icon::v1::client::xdg_toplevel_icon_v1::XdgToplevelIconV1;
+use winit_core::icon::{Icon, RgbaIcon};
+
+use crate::state::WinitState;
+
+#[derive(Debug)]
+pub struct XdgToplevelIconManagerState {
+    xdg_toplevel_icon_manager: XdgToplevelIconManagerV1,
+}
+
+impl XdgToplevelIconManagerState {
+    pub fn bind(
+        globals: &GlobalList,
+        queue_handle: &QueueHandle<WinitState>,
+    ) -> Result<Self, BindError> {
+        let xdg_toplevel_icon_manager = globals.bind(queue_handle, 1..=1, GlobalData)?;
+        Ok(Self { xdg_toplevel_icon_manager })
+    }
+
+    pub fn global(&self) -> &XdgToplevelIconManagerV1 {
+        &self.xdg_toplevel_icon_manager
+    }
+}
+
+impl Dispatch<XdgToplevelIconManagerV1, GlobalData, WinitState> for XdgToplevelIconManagerState {
+    fn event(
+        _state: &mut WinitState,
+        _proxy: &XdgToplevelIconManagerV1,
+        _event: <XdgToplevelIconManagerV1 as Proxy>::Event,
+        _data: &GlobalData,
+        _conn: &Connection,
+        _qhandle: &QueueHandle<WinitState>,
+    ) {
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum ToplevelIconError {
+    InvalidBuffer,
+    Immutable,
+    NoBuffer,
+}
+
+#[derive(Debug)]
+pub struct ToplevelIcon {
+    buffer: Buffer,
+}
+
+impl ToplevelIcon {
+    pub fn new(icon: Icon, pool: &mut SlotPool) -> Result<Self, ToplevelIconError> {
+        let buffer: Buffer;
+        let canvas: &mut [u8];
+        if let Some(icon) = icon.cast_ref::<RgbaIcon>() {
+            (buffer, canvas) = pool
+                .create_buffer(
+                    icon.width() as i32,
+                    icon.height() as i32,
+                    4 * (icon.width() as i32),
+                    Format::Argb8888,
+                )
+                .unwrap();
+
+            for (canvas_chunk, rgba) in
+                canvas.chunks_exact_mut(4).zip(icon.buffer().chunks_exact(4))
+            {
+                // Alpha in buffer is premultiplied.
+                let alpha = rgba[3] as f32 / 255.;
+                let r = (rgba[0] as f32 * alpha) as u32;
+                let g = (rgba[1] as f32 * alpha) as u32;
+                let b = (rgba[2] as f32 * alpha) as u32;
+                let color = ((rgba[3] as u32) << 24) + (r << 16) + (g << 8) + b;
+                let array: &mut [u8; 4] = canvas_chunk.try_into().unwrap();
+                *array = color.to_le_bytes();
+            }
+        } else {
+            warn!("invalid icon");
```

Comment:
Just should be an error description, no need to log a warning or our end if we can bubble up an error to the user.

Reply:
the icon can be set at two specific times:
1. at the window creation
2. at any point in time while the window exists
for 1, we can't pass the error to the user because the icon is set after the `with_window_icon` call
for 2, we can't pass the error to the user unless we change the `CoreWindow` trait
for these reasons I went with the warning, but if you meant something else can you describe it better?

Reply:
Even if it's so, I think we should just have it inside the `Error` description (which we can log into `warn!` if really needed). Also, I'm pretty sure that we have checks on the icon creation time, so it's validated to some degree, unless it's an unsupported icon, which we should log as `unsupported` indeed, invalid doesn't really tell much.